libsolidity/formal/BMC.cpp

```diff
@@ -951,18 +935,35 @@ void BMC::checkVerificationTarget(BMCVerificationTarget& _target)
 		case VerificationTargetType::Assert:
 			checkAssert(_target);
 			break;
+		case VerificationTargetType::ConstantCondition:
+			smtAssert(false, "Checks for constant condition are handled separately");
 		default:
-			solAssert(false, "");
+			smtAssert(false);
 	}
 }
 
-void BMC::checkConstantCondition(BMCVerificationTarget& _target)
+void BMC::checkIfConditionIsConstant(Expression const& _condition)
 {
+	// We ignore called functions here because they have specific input values.
+	// Also, expressions inside loop can have different values in different iterations.
+	if (!isRootFunction() || isInsideLoop())
+		return;
+
+	if (
+		!m_settings.targets.has(VerificationTargetType::ConstantCondition) ||
+		(m_currentContract && !shouldAnalyze(*m_currentContract))
+	)
+		return;
+
+	// Do not check for const-ness if this is a constant.
+	if (dynamic_cast<Literal const*>(&_condition))
```

Comment:
It looks to me like this will not exclude constants (`bool constant C = true;`), just literals (`true`), right?

```suggestion
	// Do not check for const-ness if this is a literal.
	if (dynamic_cast<Literal const*>(&_condition))
```

Reply:
I am still confused about constants in Solidity. But I believe you are right and I've updated to comment accordingly.

Reply:
Well, yeah, there's a bit of fuzziness in our terminology here. There's one syntax element we call *constant* in Solidity: a state variable with `constant` applied to it. That's what comes to mind first when is see it in a comment that refers to an AST node. But of course there are a few other things that have constant values and it's not really wrong to call them constants. Literals are one of those.

In any case, this comment definitely is referring just to literals and not constants in general.